materialize-elasticsearch/validate.go

```diff
@@ -0,0 +1,264 @@
+	constraints := make(map[string]*pm.Response_Validated_Constraint)
+	for _, projection := range boundCollection.Projections {
+		constraints[projection.Field] = validateNewProjection(projection, deltaUpdates)
+	}
+	return constraints
+}
+
+func validateMatchesExistingBinding(
+	existing *pf.MaterializationSpec_Binding,
+	boundCollection pf.CollectionSpec,
+	deltaUpdates bool,
+) map[string]*pm.Response_Validated_Constraint {
+	constraints := make(map[string]*pm.Response_Validated_Constraint)
+
+	for _, field := range existing.FieldSelection.Keys {
+		constraints[field] = &pm.Response_Validated_Constraint{
+			Type:   pm.Response_Validated_Constraint_FIELD_REQUIRED,
+			Reason: "This field is a key in the current materialization",
+		}
+	}
+
+	for _, field := range existing.FieldSelection.Values {
+		constraint := new(pm.Response_Validated_Constraint)
+
+		existingProjection := existing.Collection.GetProjection(field)
+		proposedProjection := boundCollection.GetProjection(field)
+
+		if proposedProjection == nil {
+			// A value field has been removed, so no constraint is needed.
+			continue
+		}
+
+		// Is the proposed type completely disallowed by the materialization? This differs from
+		// being UNSATISFIABLE, which implies that re-creating the materialization could resolve the
+		// difference.
+		if c := validateNewProjection(*proposedProjection, deltaUpdates); c.Type == pm.Response_Validated_Constraint_FIELD_FORBIDDEN {
+			constraints[field] = c
+			continue
+		}
+
+		existingType := propForProjection(existingProjection).Type
+		proposedType := propForProjection(proposedProjection).Type
+
+		// We match on the endpoint-specific type here and not just the JSON type to account for
+		// incompatible changes to string format fields, such as removing a numeric format. This
+		// would imply that the field is no longer a numeric value, which would be incompatible with
+		// the materialized mapping and should trigger an evolution of the materialization.
+		if existingType == proposedType {
+			constraint.Type = pm.Response_Validated_Constraint_LOCATION_RECOMMENDED
+			constraint.Reason = "This location is part of the current materialization"
+		} else {
+			constraint.Type = pm.Response_Validated_Constraint_UNSATISFIABLE
+			constraint.Reason = fmt.Sprintf(
+				"Field '%s' is already being materialized as type '%s' and cannot be changed to type '%s'",
+				field,
+				existingType,
+				proposedType,
+			)
+		}
+
+		constraints[field] = constraint
+	}
+
+	for _, proj := range boundCollection.Projections {
+		if !deltaUpdates && proj.IsRootDocumentProjection() {
+			// Standard updates materializations cannot change the projection field of the root
+			// document after the initial publication.
+			if proj.Field != existing.FieldSelection.Document {
+				constraints[proj.Field] = &pm.Response_Validated_Constraint{
+					Type: pm.Response_Validated_Constraint_UNSATISFIABLE,
```

Comment:
I think this should maybe be `FIELD_FORBIDDEN`. It's permissible for a collection to have multiple projections of the root document. And I think we'd want to say that whichever one is on the existing field selection is required, and all others are just forbidden.

I believe (but please correct me if you think otherwise) if you return an unsatisfiable constraint, then the validation crate will fail the build, even if that field isn't explicitly included.

Reply:
Interesting point about there possibly being more than one root document projection 🤔 .

I know it can get weird if a constraint is returned for a projection that doesn't exist in the collection, as would be the case if we returned a "field required" constraint for the root document projection if that projection no longer existed.

With multiple root document projections, I have observed that the first one alphabetically gets treated as the "document" when the materialization is published. The others will only be materialized as normal fields if they are explicitly required, which I guess makes since with the constraint being "location required", and one of the root projections can satisfy that. This can also get weird if the first alphabetical root projection is explicitly excluded on an initial publication, but on a later publication that exclusion is removed - then the "document" with respect to the standard updates materialization switches to the first one in alphabetical order.

I'll have to test this out and think about it some more, but I think I am tending to agree that the only strategy that makes sense is to require a single root field, but forbid the others. This would need to happen both with the initial publication (take the first one alphabetically, forbid the rest), and on later ones (require the previously seen document field, forbid other root fields).

Reply:
> This would need to happen both with the initial publication...

I don't think this would be necessary. With the initial publication, you just return `LOCATION_RECOMMENDED` for both, and the runtime will pick one. Then on later publications, we'd require the existing root projection, and forbid the others. So the user could decide whichever one they want on the initial publication, but from then on they cannot change it.

Reply:
Yep, after thinking about it for a bit more, I agree.